### PR TITLE
Feat: EIP-7702 implementation

### PR DESCRIPTION
**Pull request by mrLSD**
## Description

[EIP-7702: Set EOA account code](https://eips.ethereum.org/EIPS/eip-7702) is part of upcoming **Prague** hard fork.

➡️ Extended documentation
➡️ Code refactoring
➡️ Gas logic refactored
➡️ Access logic refactored
➡️ Authorization:
- [x] `authorization_list`
- [x] Validation for authority 
- [x] Delegation Designation  
- [x] Gas cost changes

## Tests

➡️  Refactored `evm-json-tests` for `EIP-7702`
➡️  Adde CI task for [ethereum/execution-spec-tests](https://github.com/ethereum/execution-spec-tests)
➡️  Added test for EIP-7702 from latest release [ethereum/execution-spec-tests](https://github.com/ethereum/execution-spec-tests)

### CODE REVIEW

File: gasometer/src/lib.rs [hunk 1]
```diff
@@ -574,6 +605,22 @@ pub fn static_opcode_cost(opcode: Opcode) -> Option<u32> {
...
+		if delegated_designator_is_cold == Some(true) {
+			handler.warm_target((authority_target, None));
+		}
+	}
+	let target_is_cold = handler.is_cold(target, None);
+	if target_is_cold {
+		handler.warm_target((target, None));
+	}
+	(target_is_cold, delegated_designator_is_cold)
+}
```
Comment (aleksuss) [inline gasometer/src/lib.rs:622]: More rust idiomatic would be:
```rust
fn get_and_set_warm<H: Handler>(handler: &mut H, target: H160) -> (bool, Option<bool>) {
	let delegated_designator_is_cold =
		handler
			.get_authority_target(target)
			.map(|authority_target| {
				if handler.is_cold(authority_target, None) {
					handler.warm_target((authority_target, None));
					true
				} else {
					false
				}
			});

	let target_is_cold = handler.is_cold(target, None);
	if target_is_cold {
		handler.warm_target((target, None));
	}
	(target_is_cold, delegated_designator_is_cold)
}
```

File: evm-tests/jsontests/src/utils.rs [hunk 2]
```diff
@@ -387,5 +547,40 @@ pub mod transaction {
...
 		MaxFeePerBlobGasNotSupported,
 		GasPriseEip1559,
+		AuthorizationListNotExist,
+		AuthorizationListNotSupported,
+		InvalidAuthorizationSignature,
+		CreateTransaction,
+	}
+}
+
+fn ecrecover(hash: H256, signature: &[u8]) -> Result<H160, ExitError> {
```
Comment (aleksuss) [inline evm-tests/jsontests/src/utils.rs:557]: It would be good to have a unit test for this function.

File: src/executor/stack/executor.rs [hunk 3]
```diff
@@ -1581,6 +1759,47 @@ impl<'config, 'precompiles, S: StackState<'config>, P: PrecompileSet> Handler
...
+		}
+		// Check is it loop for Delegated designation
+		self.get_authority_target(authority).map_or_else(
+			|| self.code(authority),
+			|target_address| self.code(target_address),
+		)
+	}
+
+	// Warm target according to EIP-2929
+	// It warm up the target address or storage storage. If in the target tuple
```
Comment (aleksuss) [inline src/executor/stack/executor.rs:1793]:
```suggestion
	// It warm up the target address or storage value by key. If in the target tuple
```

File: src/executor/stack/executor.rs [hunk 4]
```diff
@@ -1581,6 +1759,47 @@ impl<'config, 'precompiles, S: StackState<'config>, P: PrecompileSet> Handler
...
+	// the storage is `None` then it's warming up the address.
+	fn warm_target(&mut self, target: (H160, Option<H256>)) {
+		let (address, storage) = target;
+		if let Some(key) = storage {
+			self.state.metadata_mut().access_storage(address, key);
+		} else {
+			self.state.metadata_mut().access_address(address);
+		}
+	}
```
Comment (aleksuss) [inline src/executor/stack/executor.rs:1802]:
```rust
fn warm_target(&mut self, target: (H160, Option<H256>)) {
		match target {
			(address, None) => self.state.metadata_mut().access_address(address),
			(address, Some(key)) => self.state.metadata_mut().access_storage(address, key),
		}
	}
```

File: src/executor/stack/executor.rs [hunk 5]
```diff
@@ -1581,6 +1759,47 @@ impl<'config, 'precompiles, S: StackState<'config>, P: PrecompileSet> Handler
...
+	///
+	/// It's related to [EIP-7702 Delegation Designation](https://eips.ethereum.org/EIPS/eip-7702#delegation-designation)
+	/// When authority code is found, it should set delegated address to `authority_access` array for
+	/// calculating additional gas cost. Gas must be charged for the authority address and
+	/// for delegated address, for detection is address warm or cold.
+	fn authority_code(&mut self, authority: H160) -> Vec<u8> {
+		if !self.config.has_authorization_list {
+			return self.code(authority);
+		}
+		// Check is it loop for Delegated designation
```
Comment (aleksuss) [inline src/executor/stack/executor.rs:1785]:
```suggestion
		// Check if it is a loop for Delegated designation
```

**Review by aleksuss — COMMENTED** [review]
Overall looks good. Good job 👍🏻

File: gasometer/src/lib.rs [hunk 6]
```diff
@@ -220,9 +220,21 @@ impl<'config> Gasometer<'config> {
...
-	///
+	/// Record refund for `authority` - EIP-7702
+	/// `refunded_accounts` represent count of valid `authority`  accounts.
 	///
+	/// ## Errors
+	/// Return `ExitError` if `record_refund` operation fails.
+	pub fn record_authority_refund(&mut self, refunded_accounts: u64) -> Result<(), ExitError> {
+		let refund = i64::try_from(
+			refunded_accounts
+				* (self.config.gas_per_empty_account_cost - self.config.gas_per_auth_base_cost),
```
[the remark this answers is not in the document]
Comment (mrLSD) [reply gasometer/src/lib.rs:231]: For 32-bit overflow you need 191000 auth accounts, it's just unreachable, as you need to pay for each account before.

**Review by birchmd — APPROVED** [review]
Nice work on this!

File: gasometer/src/lib.rs [hunk 7]
```diff
@@ -409,13 +434,18 @@ pub fn call_transaction_cost(data: &[u8], access_list: &[(H160, Vec<H256>)]) ->
...
 }
 
 /// Calculate the create transaction cost.
 #[allow(clippy::naive_bytecount)]
 #[must_use]
-pub fn create_transaction_cost(data: &[u8], access_list: &[(H160, Vec<H256>)]) -> TransactionCost {
+pub fn create_transaction_cost(
+	data: &[u8],
+	access_list: &[(H160, Vec<H256>)],
+	authorization_list_len: usize,
```
Comment (aleksuss) [inline gasometer/src/lib.rs:447]: As far as I remember, we agreed that `CREATE` transactions don't have an `authorization_list`. Does it make sense to remove this parameter? Also, I can see that you pass 0 to the `record_create_transaction_cost` everywhere.

File: gasometer/src/lib.rs [hunk 8]
```diff
@@ -1208,6 +1316,8 @@ pub enum TransactionCost {
 		access_list_storage_len: usize,
 		/// Cost of initcode = 2 * ceil(len(initcode) / 32) (see EIP-3860)
 		initcode_cost: u64,
+		/// Number of authorities in transaction authorization list (see EIP-7702)
+		authorization_list_len: usize,
```
Comment (aleksuss) [inline gasometer/src/lib.rs:1320]: Redundant?

File: gasometer/src/lib.rs (hunk 7 above)
Comment (mrLSD) [reply gasometer/src/lib.rs]: Refactored: e85303d

**Review by aleksuss — APPROVED** [review]

**Pull request merged**